pandas/core/algorithms.py

```diff
@@ -2128,3 +2126,26 @@ def sort_mixed(values):
         np.putmask(new_codes, mask, na_sentinel)
 
     return ordered, ensure_platform_int(new_codes)
+
+
+def _sort_mixed(values):
+    """ order ints before strings in 1d arrays, safe in py3 """
+    str_pos = np.array([isinstance(x, str) for x in values], dtype=bool)
+    nums = np.sort(values[~str_pos])
+    strs = np.sort(values[str_pos])
+    return np.concatenate([nums, np.asarray(strs, dtype=object)])
+
+
+def _sort_tuples(values: np.ndarray[tuple]):
+    """
+    Convert array of tuples (1d) to array or array (2d).
+    We need to keep the columns separately as they contain different types and
+    nans (can't use `np.sort` as it may fail when str and nan are mixed in a
+    column as types cannot be compared).
+    """
+    from pandas.core.internals.construction import to_arrays
```

Comment:
@ssche two completely contradictory questions/requests

1) is there a reasonable way to do this without relying on core.internals?
2) could pd.MultiIndex.from_tuples de-duplicate some code by using to_arrays?

Reply:
> is there a reasonable way to do this without relying on core.internals?

My first attempt (https://github.com/pandas-dev/pandas/pull/37096/commits/1320ff12ce4f7d58dab43396df7f3d2e9e1c4792) wasn't using core.internals, as far as I recall, but it was deemed too complex.

> could pd.MultiIndex.from_tuples de-duplicate some code by using to_arrays?

Maybe, what are you trying to achieve?

Reply:
> Maybe, what are you trying to achieve?

Just simplification/de-duplication

Reply:
I don't know if this is still relevant (I forgot to respond), but I would argue that `from_arrays` should be preferred (and `from_tuples` delegating to `from_arrays`) as `from_arrays` could store type information better as it's accepting data in columnar format (instead of row-wise which `from_arrays` does).

Anyway, I hope you could proceed with your much appreciated simplification/de-dupe efforts...